custom_components/hafo/forecasters/historical_shift.py

```diff
@@ -0,0 +1,280 @@
+
+    Args:
+        hass: Home Assistant instance
+        entity_id: The sensor entity ID to fetch statistics for
+        start_time: Start of the time range
+        end_time: End of the time range
+
+    Returns:
+        List of statistics rows with 'start' and 'mean' fields.
+
+    Raises:
+        ValueError: If the recorder is not available or not set up.
+
+    """
+    if "recorder" not in hass.config.components:
+        msg = "Recorder component not loaded"
+        raise ValueError(msg)
+
+    try:
+        from homeassistant.helpers.recorder import DATA_INSTANCE  # noqa: PLC0415
+
+        if DATA_INSTANCE not in hass.data:
+            msg = "Recorder not initialized"
+            raise ValueError(msg)
+    except ImportError:
+        msg = "Recorder component not available"
+        raise ValueError(msg) from None
+
+    from homeassistant.components.recorder.statistics import statistics_during_period  # noqa: PLC0415
+
+    try:
+        statistics: dict[str, list[StatisticsRow]] = await hass.async_add_executor_job(
+            lambda: statistics_during_period(
+                hass,
+                start_time,
+                end_time,
+                {entity_id},
+                "hour",
+                None,
+                {"mean"},
+            )
+        )
+    except Exception as e:
+        msg = f"Failed to fetch statistics: {e}"
+        raise ValueError(msg) from e
+
+    return statistics.get(entity_id, [])
+
+
+def shift_history_to_forecast(
+    statistics: Sequence[StatisticsLike],
+    history_days: int,
+) -> list[ForecastPoint]:
+    """Shift historical statistics forward by N days to create a forecast.
+
+    Takes the raw hourly statistics and shifts each timestamp forward
+    by `history_days` to project them into the future.
+
+    Args:
+        statistics: List of statistics rows with 'start' and 'mean' fields.
+        history_days: Number of days to shift forward.
+
+    Returns:
+        List of ForecastPoint objects for the forecast.
+
+    """
+    forecast: list[ForecastPoint] = []
+    shift = timedelta(days=history_days)
+
+    for stat in statistics:
+        start = stat.get("start")
+        mean = stat.get("mean")
+
+        if start is None or mean is None:
+            continue
+
+        # Convert start to datetime if needed
+        if isinstance(start, datetime):
+            dt_start = start
+        else:
+            # Handle numeric timestamp (int or float)
+            try:
+                timestamp = float(start)
+                dt_start = datetime.fromtimestamp(timestamp, tz=dt_util.get_default_time_zone())
+            except (TypeError, ValueError):
+                continue
+
+        # Shift forward by N days
+        future_time = dt_start + shift
+        forecast.append(ForecastPoint(time=future_time, value=float(mean)))
+
+    # Sort by time
+    forecast.sort(key=lambda x: x.time)
+    return forecast
+
+
+def cycle_forecast_to_horizon(
```

Comment:
Haeo does this, this code shouldn't need to cycle the forecast. What we get out of the forecasting engine is the output.